[WPE][GTK] Web process cache suspend/resume does not work, WebProcessProxy::processIdentifier is not the pid of the actual web process

mcatanzaro opened the pull request:
#### c3c75352f0eef0734daa4946af7a5b60ce045ccd
<pre>
[WPE][GTK] Web process cache suspend/resume does not work, WebProcessProxy::processIdentifier is not the pid of the actual web process
<a href="https://bugs.webkit.org/show_bug.cgi?id=262794">https://bugs.webkit.org/show_bug.cgi?id=262794</a>

Reviewed by Carlos Garcia Campos.

We&apos;ll use Unix credentials to send the actual pid of the child process
to the parent process.

This could be done using the WebKit IPC connection, but the code is
much simpler if we create a separate socket for this.

* Source/WebKit/Platform/IPC/IPCUtilities.h:
* Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
(IPC::createPlatformConnection):
(IPC::sendPIDToPeer):
(IPC::readPIDFromPeer):
* Source/WebKit/Shared/unix/AuxiliaryProcessMain.cpp:
(WebKit::AuxiliaryProcessMainCommon::parseCommandLine):
* Source/WebKit/UIProcess/Launcher/glib/FlatpakLauncher.cpp:
(WebKit::flatpakSpawn):
* Source/WebKit/UIProcess/Launcher/glib/FlatpakLauncher.h:
* Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
(WebKit::ProcessLauncher::launchProcess):

Canonical link: <a href="https://commits.webkit.org/281488@main">https://commits.webkit.org/281488@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e56985f4e931717f4c5ccae6eca52c854bd8d64

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59820 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39167 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12349 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63737 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10344 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46819 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10515 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48513 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7239 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61850 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36532 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51795 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29357 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33240 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9267 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55171 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9312 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65467 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3748 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9185 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55852 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3759 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51774 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55992 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13319 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3114 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34979 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36062 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37148 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35807 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -752,7 +774,51 @@ static void addExtraPaths(const HashMap<CString, SandboxPermission>& paths, Vect
+#if HAVE(BWRAP_COMMAND_PID)
+static pid_t tryReadCommandPIDFromJSONString(CString string)
+{
+    GRefPtr<JSCContext> context = adoptGRef(jsc_context_new());
```
Maybe I should create the JSCContext once per call to readCommandPIDFromJSONStatusFile instead of again and again each time tryReadCommandPIDFromJSONString is called.

aperezdc commented on Source/WebKit/Shared/unix/AuxiliaryProcessMain.cpp:
```diff
@@ -45,18 +47,44 @@ AuxiliaryProcessMainCommon::AuxiliaryProcessMainCommon()
+    if (minimumNumArgs == 4) {
+        int pidSocket = atoi(argv[3]);
+        if (!pidSocket)
+            return false;
```
We have `WTF::parseInteger()`, which does better checking than `atoi`:

```suggestion
        auto pidSocket = parseInteger<int>(argv[3]);
        if (!pidSocket || *pidSocket < 0)
            return false;
```

mcatanzaro replied:
Good point, will change that before landing.

mcatanzaro replied:
Done.

aperezdc reviewed (commented):
Seems reasonable to me, I left a small comment.

I would prefer someone else takes a look as well before approving this, tho.

mcatanzaro commented on Source/WebKit/Platform/IPC/IPCUtilities.h:
```diff
@@ -56,9 +56,17 @@ struct SocketPair {
 enum PlatformConnectionOptions {
     SetCloexecOnClient = 1 << 0,
     SetCloexecOnServer = 1 << 1,
+#if OS(LINUX)
```
Also this guard should be `USE(GLIB) && OS(LINUX)` to match the others.

mcatanzaro commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -561,6 +561,11 @@ SocketPair createPlatformConnection(unsigned options)
             RELEASE_ASSERT_NOT_REACHED();
     }
 
+    if (options & SetPasscredOnServer) {
```
And the guards are missing here.

mcatanzaro commented on Source/WebKit/Shared/unix/AuxiliaryProcessMain.cpp:
```diff
@@ -45,18 +47,44 @@ AuxiliaryProcessMainCommon::AuxiliaryProcessMainCommon()
 {
-    ASSERT(argc >= 3);
-    if (argc < 3)
+#if OS(LINUX)
```
Should be `USE(GLIB) && OS(LINUX)`

mcatanzaro commented on Source/WebKit/Shared/unix/AuxiliaryProcessMain.cpp:
```diff
@@ -45,18 +47,44 @@ AuxiliaryProcessMainCommon::AuxiliaryProcessMainCommon()
+    if (!m_parameters.processIdentifier || !m_parameters.connectionIdentifier)
+        return false;
+
+#if OS(LINUX)
```
Should be `USE(GLIB) && OS(LINUX)`

carlosgcampos commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -554,23 +554,27 @@ SocketPair createPlatformConnection(unsigned options)
 
-        SocketPair socketPair = { sockets[0], sockets[1] };
-        return socketPair;
+        goto out;
```
Can we do the setsockopt in a helper function/lambda instead of using goto?

mcatanzaro replied:
Probably, I'll try

carlosgcampos commented on Source/WebKit/Shared/unix/AuxiliaryProcessMain.cpp:
```diff
@@ -31,11 +31,14 @@
 #endif
 
+#include "IPCUtilities.h"
+
```
This should be in the first includes block

carlosgcampos commented on Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
```diff
@@ -252,15 +263,22 @@ void ProcessLauncher::launchProcess()
+    m_processID = IPC::readPIDFromPeer(pidSocketPair.server);
+
+    RELEASE_ASSERT(!close(pidSocketPair.server));
+    RELEASE_ASSERT(!close(pidSocketPair.client));
```
Wasn't this ownership transferred to GSubprocessLauncher?

mcatanzaro replied:
Yes, that's a serious bug. Good catch. I'm surprised the assert doesn't fail.

mcatanzaro replied:
It doesn't fail because it's closed for the first time here. Later on, GSubprocessLauncher will try to close the same integer again, ignoring errors. This could close an innocent unrealted fd.

mcatanzaro committed:
[WPE][GTK] Web process cache suspend/resume does not work, WebProcessProxy::processIdentifier is not the pid of the actual web process
https://bugs.webkit.org/show_bug.cgi?id=262794

Reviewed by Carlos Garcia Campos.

We'll use Unix credentials to send the actual pid of the child process
to the parent process.

This could be done using the WebKit IPC connection, but the code is
much simpler if we create a separate socket for this.

* Source/WebKit/Platform/IPC/IPCUtilities.h:
* Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
(IPC::createPlatformConnection):
(IPC::sendPIDToPeer):
(IPC::readPIDFromPeer):
* Source/WebKit/Shared/unix/AuxiliaryProcessMain.cpp:
(WebKit::AuxiliaryProcessMainCommon::parseCommandLine):
* Source/WebKit/UIProcess/Launcher/glib/FlatpakLauncher.cpp:
(WebKit::flatpakSpawn):
* Source/WebKit/UIProcess/Launcher/glib/FlatpakLauncher.h:
* Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
(WebKit::ProcessLauncher::launchProcess):

Canonical link: https://commits.webkit.org/281488@main

neodesys commented on Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
```diff
@@ -125,10 +125,12 @@ static bool isSandboxEnabled(const ProcessLauncher::LaunchOptions& launchOptions
 void ProcessLauncher::launchProcess()
 {
-    IPC::SocketPair socketPair = IPC::createPlatformConnection(connectionOptions());
+    RELEASE_ASSERT(m_launchOptions.processType != ProcessLauncher::ProcessType::DBusProxy);
```
This is breaking the build if `BUBBLEWRAP_SANDBOX` is disabled because `ProcessLauncher::ProcessType::DBusProxy` is only defined if the sandbox is enabled.

Should be something like this, maybe:
```
#if ENABLE(BUBBLEWRAP_SANDBOX)
    RELEASE_ASSERT(m_launchOptions.processType != ProcessLauncher::ProcessType::DBusProxy);
#endif
```

mcatanzaro replied:
Yes, fix here: https://github.com/WebKit/WebKit/pull/31461

mcatanzaro replied:
BTW, if you're working on the Android support, I wonder if it might be simpler to have a completely separate ProcessLauncher for libwpe on non-Linux? I'm not sure. It's starting to feel like a lot of conditionals, though. :)
